File: Frontend/workspace-2a3ac13c-ec98-4e6d-b282-2b37ac9303e6/src/types/article.ts
// Types matching the BHUB Python backend schemas

export interface Author {
  id: number;
  name: string;
  orcid?: string | null;
  affiliation?: string | null;
  avatar?: string;
  role?: 'author' | 'editor';
  article_count?: number;
}

export interface Category {
  id: number;
  name: string;
  slug: string;
  description?: string | null;
  color: string;
  keywords?: string | null;
  article_count?: number;
  created_at?: string;
  updated_at?: string;
}

export interface Article {
  id: number;
  title: string;
  abstract?: string | null;
  abstract_translated?: string | null;
  title_translated?: string | null;
  keywords?: string | null;
  original_url?: string | null;
  pdf_url?: string | null;
  image_url?: string | null;
  publication_date?: string | null;
  journal_name?: string | null;
  volume?: string | null;
  issue?: string | null;
  pages?: string | null;
  doi?: string | null;
  language: string;
  external_id?: string | null;
  category_id?: number | null;
  category?: Category | null;
  impact_score: number;
  classification_confidence?: number | null;
  highlighted: boolean;
  is_published: boolean;
  source_type: string;
  source_category?: 'journal' | 'portal';
  feed_id?: number | null;
  feed_name?: string | null;
  pdf_file_path?: string | null;
  pdf_file_size?: number | null;
  view_count: number;
  download_count: number;
  authors: Author[];
  has_pdf: boolean;
  created_at?: string;
  updated_at?: string;
}

// Paginated response from backend
export interface PaginatedArticleResponse {
  items: Article[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

// Category list response from backend
export interface CategoryListResponse {
  categories: Category[];
}

// Similar articles response
export interface ArticleSimilarResponse {
  articles: Article[];
}

// Frontend-specific props for components
export interface ArticleCardProps {
  id: number;
  title: string;
  excerpt: string;
  category: string;
  author: {
    name: string;
    avatar?: string;
  };
  authors?: Author[]; // Added authors support
  date: string;
  readingTime: number;
  source?: string;
  sourceId?: number;
  sourceCategory?: 'journal' | 'portal';
  imageUrl?: string | null;
  featured?: boolean;
  variant?: 'default' | 'compact';
}

export interface FeaturedArticleCardProps {
  id: number;
  title: string;
  excerpt: string;
  category: {
    label: string;
    icon?: React.ReactNode;
  };
  date: string;
  citations: number;
  author: {
    name: string;
    avatar: string;
    affiliation: string;
  };
  readingTime: number;
  onRead?: () => void;
  onLike?: () => void;
  isLiked?: boolean;
}

export interface ListParams {
  category_id?: number;
  author?: string;
  search?: string;
  page?: number;
  limit?: number;
  sort_by?: 'publication_date' | 'title' | 'impact_score' | 'view_count' | 'created_at';
  sort_order?: 'asc' | 'desc';
  highlighted?: boolean;
  has_pdf?: boolean;
  date_from?: string;
  date_to?: string;
  strategy?: 'default' | 'interleaved';
  source_category?: 'journal' | 'portal';
  feed_id?: number;
}

// Helper functions to transform backend data to frontend component props
export function articleToCardProps(article: Article): ArticleCardProps {
  return {
    id: article.id,
    title: article.title,
    excerpt: article.abstract?.substring(0, 200) || '',
    category: article.category?.name || 'Uncategorized',
    author: {
      name: article.authors[0]?.name || 'Unknown Author',
      avatar: undefined,
    },
    authors: article.authors, // Pass full authors list
    date: article.publication_date || article.created_at || '',
    readingTime: Math.ceil((article.abstract?.length || 0) / 1000) || 5,
    source: article.feed_name || article.journal_name || undefined,
    sourceId: article.feed_id || undefined,
    sourceCategory: article.source_category as 'journal' | 'portal' | undefined,
    imageUrl: article.image_url,
    featured: article.highlighted,
  };
}

export function articleToFeaturedProps(article: Article): Omit<FeaturedArticleCardProps, 'onRead' | 'onLike' | 'isLiked'> {
  return {
    id: article.id,
    title: article.title,
    excerpt: article.abstract?.substring(0, 300) || '',
    category: {
      label: article.category?.name || 'Uncategorized',
    },
    date: article.publication_date || article.created_at || '',
    citations: article.view_count,
    author: {
      name: article.authors[0]?.name || 'Unknown Author',
      avatar: '',
      affiliation: article.authors[0]?.affiliation || '',
    },
    readingTime: Math.ceil((article.abstract?.length || 0) / 1000) || 5,
  };
}

File: bhub-backend-python/app/schemas/article.py
"""
Schemas de artigo.
"""

from datetime import datetime

from pydantic import Field, HttpUrl, computed_field

from app.models.article import SourceType
from app.schemas.category import CategoryResponse
from app.schemas.common import BaseSchema, PaginatedResponse, TimestampSchema


class AuthorResponse(BaseSchema):
    """Resposta de autor."""

    id: int
    name: str
    orcid: str | None = None
    affiliation: str | None = None
    role: str | None = "author"


class ArticleBase(BaseSchema):
    """Campos base de artigo."""

    title: str = Field(..., min_length=1, max_length=500)
    abstract: str | None = None
    keywords: str | None = None
    original_url: str | None = Field(default=None, max_length=500)
    pdf_url: str | None = Field(default=None, max_length=500)
    image_url: str | None = Field(default=None, max_length=500)
    publication_date: datetime | None = None
    journal_name: str | None = Field(default=None, max_length=255)
    volume: str | None = Field(default=None, max_length=50)
    issue: str | None = Field(default=None, max_length=50)
    pages: str | None = Field(default=None, max_length=50)
    doi: str | None = Field(default=None, max_length=100)
    language: str = "en"


class ArticleCreate(ArticleBase):
    """Schema de criação de artigo."""

    category_id: int | None = None
    feed_id: int | None = None
    authors: list[str] = []  # Lista de nomes de autores


class ArticleUpdate(BaseSchema):
    """Schema de atualização de artigo."""

    title: str | None = Field(default=None, min_length=1, max_length=500)
    abstract: str | None = None
    keywords: str | None = None
    original_url: str | None = None
    pdf_url: str | None = None
    image_url: str | None = None
    publication_date: datetime | None = None
    journal_name: str | None = None
    category_id: int | None = None
    highlighted: bool | None = None
    impact_score: float | None = Field(default=None, ge=1, le=10)
    is_published: bool | None = None


class ArticleResponse(ArticleBase, TimestampSchema):
    """Resposta de artigo."""

    id: int
    external_id: str | None = None
    title_translated: str | None = None
    abstract_translated: str | None = None
    category_id: int | None = None
    category: CategoryResponse | None = None
    impact_score: float
    classification_confidence: float | None = None
    highlighted: bool
    is_published: bool
    source_type: SourceType
    feed_id: int | None = None
    feed_name: str | None = None
    pdf_file_path: str | None = None
    pdf_file_size: int | None = None
    view_count: int
    download_count: int
    authors: list[AuthorResponse] = []
    has_pdf: bool = False
    has_pdf: bool = False

    @computed_field
    def source_category(self) -> str:
        """Categoria da fonte (computed)."""
        # Journals are PDF/MANUAL OR have a journal_name defined (even if RSS)
        if self.source_type in [SourceType.PDF, SourceType.MANUAL] or self.journal_name:
            return "journal"
        return "portal"


class ArticleListResponse(PaginatedResponse[ArticleResponse]):
    """Lista paginada de artigos."""

    pass


class ArticleSearchParams(BaseSchema):
    """Parâmetros de busca de artigos."""

    search: str | None = Field(default=None, min_length=2, max_length=200)
    category_id: int | None = None
    author: str | None = None
    feed_id: int | None = None
    highlighted: bool | None = None
    has_pdf: bool | None = None
    date_from: datetime | None = None
    date_to: datetime | None = None
    sort_by: str = Field(default="publication_date", pattern=r"^(publication_date|title|impact_score|view_count|created_at)$")
    sort_order: str = Field(default="desc", pattern=r"^(asc|desc)$")


class ArticleSimilarResponse(BaseSchema):
    """Artigos similares."""

    articles: list[ArticleResponse]


class ArticleHighlightRequest(BaseSchema):
    """Request para destacar artigo."""

    highlighted: bool


class ScrapeRequest(BaseSchema):
    """Request para scraping de URL."""

    url: str = Field(..., max_length=500)
    category_id: int | None = None


class ScrapeResponse(BaseSchema):
    """Resposta de scraping."""

    success: bool
    article: ArticleResponse | None = None
    error: str | None = None


class PDFUploadResponse(BaseSchema):
    """Resposta de upload de PDF."""

    success: bool
    article_id: int | None = None
    message: str
    duplicate: bool = False
